pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/NullHandlingIntegrationTest.java

```diff
@@ -219,97 +213,187 @@ public void testTotalCountWithNullHandlingQueryOptionEnabled(boolean useMultiSta
     assertTrue(response.get("exceptions").isEmpty());
-    assertEquals(rows.size(), 1);
-    assertEquals(rows.get(0).get(0).asBoolean(), false);
+    assertEquals(1, rows.size());
+    assertFalse(rows.get(0).get(0).asBoolean());
 
     sqlQuery = "SELECT isNotDistinctFrom(null, null) FROM " + getTableName() + "  OPTION (enableNullHandling=true);";
     response = postQuery(sqlQuery);
     rows = response.get("resultTable").get("rows");
     assertTrue(response.get("exceptions").isEmpty());
-    assertEquals(rows.size(), 1);
-    assertEquals(rows.get(0).get(0).asBoolean(), true);
+    assertEquals(1, rows.size());
+    assertTrue(rows.get(0).get(0).asBoolean());
 
 
     sqlQuery = "SELECT isDistinctFrom(null, 1) FROM " + getTableName() + "  OPTION (enableNullHandling=true);";
     response = postQuery(sqlQuery);
     rows = response.get("resultTable").get("rows");
     assertTrue(response.get("exceptions").isEmpty());
-    assertEquals(rows.size(), 1);
-    assertEquals(rows.get(0).get(0).asBoolean(), true);
+    assertEquals(1, rows.size());
+    assertTrue(rows.get(0).get(0).asBoolean());
 
     sqlQuery = "SELECT isNotDistinctFrom(null, 1) FROM " + getTableName() + "  OPTION (enableNullHandling=true);";
     response = postQuery(sqlQuery);
     rows = response.get("resultTable").get("rows");
     assertTrue(response.get("exceptions").isEmpty());
-    assertEquals(rows.size(), 1);
-    assertEquals(rows.get(0).get(0).asBoolean(), false);
+    assertEquals(1, rows.size());
+    assertFalse(rows.get(0).get(0).asBoolean());
 
     sqlQuery = "SELECT case when true then null end FROM " + getTableName() + "  OPTION (enableNullHandling=true);";
     response = postQuery(sqlQuery);
     rows = response.get("resultTable").get("rows");
     assertTrue(response.get("exceptions").isEmpty());
-    assertEquals(rows.size(), 1);
-    assertEquals(rows.get(0).get(0).asText(), "null");
+    assertEquals(1, rows.size());
+    assertEquals("null", rows.get(0).get(0).asText());
 
 
     sqlQuery = "SELECT case when false then 1 end FROM " + getTableName() + "  OPTION (enableNullHandling=true);";
     response = postQuery(sqlQuery);
     rows = response.get("resultTable").get("rows");
     assertTrue(response.get("exceptions").isEmpty());
-    assertEquals(rows.size(), 1);
-    assertEquals(rows.get(0).get(0).asText(), "null");
+    assertEquals(1, rows.size());
+    assertEquals("null", rows.get(0).get(0).asText());
 
 
     // Null intolerant functions
     sqlQuery = "SELECT add(null, 1) FROM " + getTableName() + "  OPTION (enableNullHandling=true);";
     response = postQuery(sqlQuery);
     rows = response.get("resultTable").get("rows");
     assertTrue(response.get("exceptions").isEmpty());
-    assertEquals(rows.size(), 1);
-    assertEquals(rows.get(0).get(0).asText(), "null");
+    assertEquals(1, rows.size());
+    assertEquals("null", rows.get(0).get(0).asText());
+  }
+
+  @Test
+  public void testNullLiteralSelectionInV2() throws Exception {
```

Comment:
Can we rewrite these tests as one method per test or, maybe better, one provider with the query and the expected results?

We have tons of tests like this (or the one above) where the test has a lot of assertions. That is an anti-pattern we should try to get rid of. We never have time to modify the old code, but we should try to write new tests in a better way.

Just to be clear, there are two main advantages of the suggested approach:
1. The tests are easier to read (each test is a query and an expected result)
2. In case some test fail, the following tests are executed. In the current version if test 2 fails, 3, 4, etc are not executed, which means it can be more difficult to find the actual problem (or to know if there are more than one!)

Reply:
Makes sense, I've added a data provider and updated both these tests. Let me know if this is along the lines of what you were thinking about. We could very easily add another parameter for v1 / v2 and use that to determine the query engine and the expected number of rows in order to combine both the test methods into a single one. However, given the notable difference in behavior between the two engines in this case (v1 simply handles it in the broker and returns a single result), I preferred keeping them as separate tests using the same data provider to emphasize the difference.